// @colony/purser-trezor/messages.js
"use strict";
/* @flow */
/* eslint-disable max-len */
Object.defineProperty(exports, "__esModule", { value: true });
exports.classInstance = {
    signContractDeployment: 'In order to sign a contract deployment transaction (Not having a destination address), the Trezor service requires the `inputData` prop value to be set',
};
exports.staticMethods = {
    userExportCancel: 'User cancelled the account export request (via Window prompt)',
    userExportGenericError: 'Could not export the wallet account, check the values you are sending to the Trezor service',
    userSignTxCancel: 'User cancelled signing the transaction (via Hardware buttons)',
    userSignTxGenericError: 'Could not sign the transaction, check the values you are sending to the Trezor service',
    messageSignatureInvalid: 'The message signature is invalid',
    messageSignatureOnlyTrezor: 'Please take note: The message signature produced by a trezor wallet can only be verified using that trezor wallet (verifyMessage method)',
};
